### PR TITLE
Sort heatmap

### PR DESCRIPTION
**Pull request by hakkd**
implementing hierarchical clustering and optimal leaf ordering for heatmap data to show nutritionally-similar foods.

### CODE REVIEW

**Comment by hakkd** [comment]
@joelostblom I implemented a sorting function to sort the heatmap using linkage. This seems to work pretty well. I also quickly fixed the overlapping dropdown selectors bug by changing to a the bootstrap panel template. The general layout is unchanged, but I figured we can refine the design later anyway.

File: nutrimap.py [hunk 1]
```diff
@@ -249,7 +252,7 @@ def get_food_group(food) -> str:
 
 # create a scatter plot filtered by food/nutrient group reduced to 2 dimensions
 def pca_scatter_2_components(data):
-    data = pd.pivot(data, index="food", columns="nutrient").reset_index().fillna(0)
+    data = pd.pivot(data, index="food", columns="nutrient", values="rdi").reset_index().fillna(0)
```
Comment (joelostblom) [inline nutrimap.py:255]: Although filling NA values with `0` can seem harmless, it is actually a quite large assumption to say that if a value is missing, it was probably 0. E.g. all the values might be high for other foods for a certain feature, and if we introduce a 0, then this food item will appear as that we know it is very different from the rest, when in reality we just don't know what the value is for a specific feature. A less strong assumption is usually to use a more advanced way of filling the nulls, e.g. via interpolation or a predictive method using the information in the other columns. In our case, we have an advantage because we can find out the actual values by looking them up, so I think we should do that as often as possible, and as a fallback use the mean of the column, because that it less likely to introduce difference on average to the other rows. This change will affect both PCA and clustering (heatmap sorting) below (it should still display as NA in the heatmap thought, this change is just for the ordering and the scatter plot).

I have these values since previously which we can use for the most important missing values ([from here](https://github.com/joelostblom/nutrimap/blob/main/old-bokeh-version/nutrimap/preprocessing.ipynb)):

```
foods_sub.loc['Oats', 'Sugars'] = 1.1
foods_sub.loc['Oats', 'Selenium'] = 28.9
foods_sub.loc['Oats', 'Vitamin E'] = 0.42
foods_sub.loc['Oats', 'Vitamin K'] = 2
foods_sub.loc['Oats', 'beta-Carotene'] = 0
foods_sub.loc['Oats', 'alpha-Carotene'] = 0
foods_sub.loc['Oats', 'beta-Cryptoxanthin'] = 0
foods_sub.loc['Oats', 'Lycopene'] = 0
foods_sub.loc['Oats', 'Lutein + zeaxanthin'] = 180

foods_sub.loc['Quinoa, uncooked', 'Sugars'] = 6.1
foods_sub.loc['Quinoa, uncooked', 'Vitamin C'] = 0
foods_sub.loc['Buckwheat', 'Sugars'] = 1.9
foods_sub.loc['Millet, raw', 'Sugars'] = 1.5
```

Could you add this code and modify it as needed? As well as setting the mean as the default value for all columns when there is missing NA values?

File: nutrimap.py [hunk 2]
```diff
@@ -289,6 +292,22 @@ def pca_scatter_2_components(data):
     
     return chart
 
+# sort data using hierarchical clustering and optimal leaf-ordering
+def sort_similar_foods(filtered_df):
+    """
+    requires that the data matches the input of create_heatmap function
+    """
+    wide_data = pd.pivot(filtered_df, index="food", columns="nutrient", values="rdi").reset_index().fillna(0)
```
Comment (joelostblom) [inline nutrimap.py:300]: Same comment here about filling NAs as above

File: nutrimap.py [hunk 3]
```diff
@@ -289,6 +292,22 @@ def pca_scatter_2_components(data):
...
+    wide_data.columns = wide_data.columns.get_level_values(0)
+
+    X = wide_data.iloc[:, 1:]
+
+    Z = hierarchy.linkage(X, optimal_ordering=True)
```
Comment (joelostblom) [inline nutrimap.py:305]: The default linkage method in `linkage` is `'single'`, which mean look at the distance of the closest values between the rows (so a single column). I don't think that aligns with the task we are trying to achieve which is to group rows that appear visually similar in the heatmap together. To me this would be something like the average of the distances between all rows, so I think `'average'` is a good value for this ('ward' could also work, but 'average' looked better when I played around with it, some useful intuition about the different metrics [can be found here](https://stats.stackexchange.com/questions/195446/choosing-the-right-linkage-method-for-hierarchical-clustering)).

File: nutrimap.py [hunk 4]
```diff
@@ -289,6 +292,22 @@ def pca_scatter_2_components(data):
...
+
+    # find the optimal order of row indexes according to the clustering algorithm
+    optimal_order = hierarchy.leaves_list(Z)
+
+    return optimal_order
```
Comment (joelostblom) [inline nutrimap.py:310]: As your comment has noted, this is returning an array of row indexes. However, passing this list directly to the `sort` option in altair will produce a result that does not appear sorted:

image

In the heatmap, `'corn'` definitely looks out of place and it doesn't fit with our notion of similarity between rows. It is often useful to perform a visual inspection like this to see if our results make sense. In this case the issue is that you are passing row indices of the data frame to `sort` while you actually need to pass the order of the values on the y-axis (so the food item names) via e.g. `wide_data.loc[optimal_order, 'food'].tolist()` (note that in this case the dataframes are also of different shape so even if altair would accept row indices they would not have lined up). If we make this change, we get this heatmap instead which appear sorted closer to what we would expect:

image

File: nutrimap.py [hunk 5]
```diff
@@ -337,7 +356,7 @@ def make_plot(food_group, nutrient_group, max_dv):
     return alt.vconcat(scatter, heatmap)
 
 # Build the dashboard
-pn.template.FastListTemplate(
+pn.template.BootstrapTemplate(
```
Comment (joelostblom) [inline nutrimap.py:359]: I don't think we need to make this change. After merging #12 which updated panel, the z-order should no longer be an issue, just make sure you reinstalled and updated panel locally. Having that said, if there is something else you prefer in the Bootstrap template, I'm happy to keep the change.

File: nutrimap.py [hunk 6]
```diff
@@ -223,7 +226,7 @@ def compute_rdi_proportion(row):
 food_group = pn.widgets.MultiChoice(
     name='Food Groups',
     value=['vegetables', 'grains'],
-    options=list(food_groups.keys()),
+    options=list(food_groups.keys())
 )
```
Comment (joelostblom) [inline nutrimap.py:230]: Minor note on this: The trailing comma can be nice because that means if you add a new item, only the line of the new item will be marked as changed, since the comma already existed on the previous line. I don't really care which approach we use here as long as we are consistent throughout the project, just a footnote that this is sometimes done intentionally.

**Review by joelostblom — CHANGES_REQUESTED** [review]
Looking great overall and almost there @hakkd ! I have a few suggested changes that you can work on and then ping me again for review.

File: nutrimap.py [hunk 7]
```diff
@@ -4,6 +4,9 @@
 import numpy as np
 from sklearn.preprocessing import StandardScaler
 from sklearn.decomposition import PCA as pca
+from scipy.cluster import hierarchy 
+from scipy.cluster.hierarchy import ward, median, centroid, weighted, average, complete, single, fcluster
+from scipy.spatial.distance import pdist
```
[the remark this answers is not in the document]
Comment (hakkd) [reply nutrimap.py:9]: I was testing some of the other clustering methods like Ward's but forgot to delete the imports

File: nutrimap.py (hunk 5 above)
Comment (hakkd) [reply nutrimap.py:389]: After updating my panel version and trying FastListTemplate there was some strange behaviour with the component positioning. I've kept the Bootstrap template for now.

File: nutrimap.py (hunk 1 above)
Comment (hakkd) [reply nutrimap.py]: Should I save that dataframe with the missing values added?

File: nutrimap.py (hunk 5 above)
Comment (joelostblom) [reply nutrimap.py:389]: Sounds good

File: nutrimap.py (hunk 1 above)
Comment (joelostblom) [reply nutrimap.py]: You don't need to save it to disk for now

**Review by joelostblom — APPROVED** [review]
Looks great, thanks @hakkd !
